### PR TITLE
DSL ActorContext extends typed ActorContext #24823

### PR DESCRIPTION
**Pull request by rkuhn**

### CODE REVIEW

**Comment by rkuhn** [comment]
fixes #24823

**Review by chbatey — APPROVED** [review]
LGTM, i don't see a downside to this unless we want to make interpretMessage internal

**Review by patriknw — CHANGES_REQUESTED** [review]
I think this is a good change.

This will expose `asJava` in javadsl and `asScala` scaladsl, but that's not too bad.

A few internal casts can now be removed, search for "asInstanceOf.*ActorContext" in akka-actor-typed

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/BehaviorImpl.scala [hunk 1]
```diff
@@ -27,10 +27,9 @@ import scala.reflect.ClassTag
 
   def widened[T, U](behavior: Behavior[T], matcher: PartialFunction[U, T]): Behavior[U] = {
     behavior match {
-      case d: DeferredBehavior[T] ⇒
+      case d: DeferredBehavior[t] ⇒
         DeferredBehavior[U] { ctx ⇒
-          val c = ctx.asInstanceOf[akka.actor.typed.ActorContext[T]]
-          val b = Behavior.validateAsInitial(Behavior.start(d, c))
+          val b = Behavior.validateAsInitial(Behavior.start(d, ctx.as[t]))
```
Comment (ktoso) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/BehaviorImpl.scala]: How could I have missed that we have such `as[]`, was missing it a few times

**Review by patriknw — APPROVED** [review]
LGTM, thanks

**Pull request merged**